docs_nnx/guides/transforms.md

```diff
@@ -0,0 +1,340 @@
+
+| Lift Type        | Transforms                              |
+|------------------|-----------------------------------------|
+| `StateAxes`      | `vmap`, `pmap`, `scan`                  |
+| `StateSharding`  | `jit`, `shard_map`                      |
+| `DiffState`      | `grad`, `value_and_grad`, `custom_vjp`  |
+
+> NOTE: `shard_map` is not yet implemented.
+
+If we want to specify how to vectorize different substates of an object in `nnx.vmap`, we create a `StateAxes` which maps a set of substates via [Filters](https://flax-nnx.readthedocs.io/en/latest/guides/filters_guide.html) to their corresponding axes, and pass the `StateAxes` to `in_axes` and `out_axes` as if it were a pytree prefix. Let's use the previous `stateful_vector_dot` example and
+vectorize only the `Param` variables and broadcast the `count` variable so we only keep a single count for all the batch elements.
+To do this we will define a `StateAxes` with a filter that matches the `Param` variables and maps them to axis `0`, and all the `Count` variables to `None`, and pass this `StateAxes` to `in_axes` for the `Weights` object.
+
+```{code-cell} ipython3
+class Weights(nnx.Module):
+  def __init__(self, kernel: jax.Array, bias: jax.Array, count: jax.Array):
+    self.kernel, self.bias = nnx.Param(kernel), nnx.Param(bias)
+    self.count = Count(count)
+
+weights = Weights(
+  kernel=random.uniform(random.key(0), (10, 2, 3)),
+  bias=jnp.zeros((10, 3)),
+  count=jnp.array(0),
+)
+x = jax.random.normal(random.key(1), (10, 2))
+
+
+def stateful_vector_dot(weights: Weights, x: jax.Array):
+  assert weights.kernel.ndim == 2, 'Batch dimensions not allowed'
+  assert x.ndim == 1, 'Batch dimensions not allowed'
+  weights.count += 1
+  return x @ weights.kernel + weights.bias
+
+state_axes = nnx.StateAxes({nnx.Param: 0, Count: None}) # broadcast Count
+y = nnx.vmap(stateful_vector_dot, in_axes=(state_axes, 0), out_axes=1)(weights, x)
+y = nnx.vmap(stateful_vector_dot, in_axes=(state_axes, 0), out_axes=1)(weights, x)
+
+weights.count
+```
+
+Here count is now a scalar since its not being vectorized. Also, note that `StateAxes` can only be used directly on Flax NNX objects, it cannot be used as a prefix for a pytree of objects.
+
++++
+
+### Random State
+In Flax NNX random state is just regular state. This means that its stored inside Modules that need it and its treated as any other type of state. This is a simplification over Flax Linen where random state was handled by a separate mechanism. In practice Modules usually keep that need random state simply need a references to a `Rngs` object that is passed to them during initialization, and
```

Comment:
> In practice Modules usually keep that need random state simply need a references to a Rngs object that is passed to them during initialization, and use it to generate a unique key for each random operation.

What about:
In practice Modules simply need to keep a reference to a Rngs object that is passed to them during initialization, and use it to generate a unique key for each random operation.